### PR TITLE
p2p/conn: Add Bufferpool

### PR DESCRIPTION
**Pull request by guagualvcha**
According to benchmark, the read/write performance improved:

**Read** : 
Before:
BenchmarkReadSecretConnection-4   	  300000	      4747 ns/op	    4672 B/op	       6 allocs/op
After
BenchmarkReadSecretConnection-4   	  300000	      3896 ns/op	       0 B/op	       0 allocs/op

**Write**:
Before:
BenchmarkWriteSecretConnection-4   	  200000	      8477 ns/op	    8006 B/op	      10 allocs/op
After:
BenchmarkWriteSecretConnection-4   	  200000	      6765 ns/op	       0 B/op	       0 allocs/op

We test this with 11 validators:

we do not see obvious improvement in TPS,  but the go gc decrease from 30-25 to 25-10 when we run 250 tps.
 
* [x] Updated all relevant documentation in docs
* [x] Updated all code comments where relevant
* [x] Wrote tests
* [ ] Updated CHANGELOG_PENDING.md

### CODE REVIEW

File: p2p/conn/secret_connection.go [hunk 1]
```diff
@@ -143,35 +154,37 @@ func (sc *SecretConnection) Write(data []byte) (n int, err error) {
...
-		aead.Seal(sealedFrame[:0], sc.sendNonce[:], frame, nil)
-		incrNonce(sc.sendNonce)
-		// end encryption
-
-		_, err = sc.conn.Write(sealedFrame)
-		if err != nil {
+		if err := func() error {
+			var sealedFrame = pool.Get(aeadSizeOverhead + totalFrameSize)
+			var frame = pool.Get(totalFrameSize)
+			defer pool.Put(sealedFrame)
```
Comment (melekes) [inline p2p/conn/secret_connection.go]: can we unify these

```go
defer func() {
}
```

**Comment by melekes** [comment]
from https://github.com/libp2p/go-buffer-pool#disadvantages-over-gc
> Can leak memory contents. Unlike the go GC, we don't zero memory.

This is not a concern here, right?

**Comment by guagualvcha** [comment]
> from https://github.com/libp2p/go-buffer-pool#disadvantages-over-gc
> 
> > Can leak memory contents. Unlike the go GC, we don't zero memory.
> 
> This is not a concern here, right?

I think so.  The buffer length we use is not dynamic, but two fixed length.

File: p2p/conn/secret_connection.go (hunk 1 above)
Comment (guagualvcha) [reply p2p/conn/secret_connection.go]: changed.

**Comment by marbar3778** [comment]
@guagualvcha could you resolve the merge conflict here?

**Comment by guagualvcha** [comment]
> @guagualvcha could you resolve the merge conflict here?

resolved

**Comment by marbar3778** [comment]
Thank you! Could the gopkg.toml and .lock be removed? These are not being used anymore in the repo.

**Comment by guagualvcha** [comment]
> Thank you! Could the gopkg.toml and .lock be removed? These are not being used anymore in the repo.

done

**Comment by melekes** [comment]
Would be great if we could add this with a feature flag. I am a little bit concerned about adding yet another dependency. But gains in GC could be worth it.

**Comment by marbar3778** [comment]
I don't think we should be concerned with the added dependency here as we are starting to explore libp2p as an alternative to our current p2p stack and will have to add many of the libraries from libp2p as dependencies. We also don't want to start throwing many things under feature flags, IMO. Only when needed

**Comment by guagualvcha** [comment]
hi @marbar3778 @melekes  is this pr ready for merge?

**Comment by marbar3778** [comment]
sorry for the delay, could you add an entry into the changelog under improvements then I will merge

**Comment by guagualvcha** [comment]
sure, will do

**Pull request merged**